### PR TITLE
Use JSON for favorites, move server list code to Lua

### LINKED ISSUE #5878: Favorite server does not get “unstarred” when deleting favorite with active search
How to reproduce:

* Use a7787bb9d27851b3e8ff9ff388e50793ee5120ed
* Go to tab “Play Online”
* Enter something into search, so that from your favorite servers (those with a star), one or more is still visible
* Select one of your favorite servers
* Click “Delete favorite”

Expected: Server gets unfavorited and unstarred
Actual: Server keeps its star icon. But when you empty the search bar, the server is gone from the favourites.

Compare this behavour to the behaviour when you have an empty search.

### LINKED ISSUE #9933: Can’t favourite multiple unlisted servers
##### Minetest version
```
Minetest 5.2.0 (Linux)
Using Irrlicht 1.8.4
BUILD_TYPE=Gentoo
RUN_IN_PLACE=0
USE_GETTEXT=1
USE_SOUND=1
USE_CURL=1
USE_FREETYPE=1
USE_LUAJIT=1
STATIC_SHAREDIR="/usr/share/minetest"
```

##### OS / Hardware
Operating system: Linux
CPU: Intel(R) Core(TM) i7-3770 CPU @ 3.40GHz

##### Summary
I connected to an unlisted server by hostname and port number. It became a favourite and was shown at the top of the server list with a gold star. I then connected to a second unlisted server. Although both servers appear in `serverlist/favoriteservers.txt`, only the most recent one actually shows up in the GUI; the older one is not visible.

##### Steps to reproduce
Connect to two unlisted servers in a row, by hostname and port number. Go back to the server list in the GUI.

### PR DESCRIPTION
**Pull request by rubenwardy**
- Favourites are now serialised using JSON. The old gross text format is supported for reading only.
- Move server list code to Lua
- This introduces a slight change in behaviour: joining a server using the command line won't add a favorite. I don't think this is a problem though.
- Fixes #5878  Fixes #9933

## To do

- [x] Set favorite on join server
- [x] Fix online server list
- [x] Remove favorites in `menudata`
- [x] Clean up serverlistmgr API

### CODE REVIEW

**Comment by nerzhul** [comment]
i really like this change

**Comment by LoneWolfHT** [comment]
tab_online.lua:110: Attempt to index a nil value

**Comment by LoneWolfHT** [comment]
I see multiple entries for local host in my serverlist.json
Seems like one is added for every join?

**Comment by rubenwardy** [comment]
Fixed. That crash was due to string `"30000"` being compared to number `30000`

File: builtin/mainmenu/serverlistmgr.lua [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+function serverlistmgr.get_favorites()
+	if serverlistmgr.favorites then
+		return serverlistmgr.favorites
+	end
+
+	serverlistmgr.favorites = {}
+
+	-- Add favourites, removing duplicates
+	local seen = {}
+	for _, fav in ipairs(read_favorites()) do
```
Comment (rubenwardy) [inline builtin/mainmenu/serverlistmgr.lua:200]:
```suggestion
	for _, fav in ipairs(read_favorites() or {}) do
```

File: builtin/mainmenu/serverlistmgr.lua [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+	local path = get_favorites_path()
+	core.create_dir(path)
+	core.safe_file_write(path, core.write_json(favorites))
+end
+
+--------------------------------------------------------------------------------
+function serverlistmgr.read_legacy_favorites(path)
+	local file = io.open(path, "r")
+	if file then
```
Comment (rubenwardy) [inline builtin/mainmenu/serverlistmgr.lua:104]: invert if-statement

File: builtin/mainmenu/serverlistmgr.lua [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+		core.settings:set("serverlist_file", filename:sub(1, #filename - 4) .. ".json")
+	end
+
+	local path = get_favorites_path()
+	core.create_dir(path)
+	core.safe_file_write(path, core.write_json(favorites))
+end
+
+--------------------------------------------------------------------------------
+function serverlistmgr.read_legacy_favorites(path)
```
Comment (rubenwardy) [inline builtin/mainmenu/serverlistmgr.lua:102]: Should this delete the legacy file? I guess it's cleaner to do so, and favourites aren't exactly vital to keep a backup of

File: builtin/mainmenu/serverlistmgr.lua [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+
+--------------------------------------------------------------------------------
+function serverlistmgr.get_favorites()
+	if serverlistmgr.favorites then
+		return serverlistmgr.favorites
+	end
+
+	serverlistmgr.favorites = {}
+
+	-- Add favourites, removing duplicates
```
Comment (rubenwardy) [inline builtin/mainmenu/serverlistmgr.lua:198]: Someone watch out for this :gb: 

```suggestion
	-- Add favorites, removing duplicates
```

File: builtin/mainmenu/serverlistmgr.lua (hunk 3 above)
Comment (LoneWolfHT) [reply builtin/mainmenu/serverlistmgr.lua:102]: Sounds fine to me

File: minetest.conf.example [hunk 5]
```diff
@@ -1928,7 +1928,7 @@
 #    octaves     = 3,
 #    persistence = 0.5,
 #    lacunarity  = 2.0,
-#    flags       =
```
Comment (sfan5) [inline minetest.conf.example:1931]: needs undoing

**Review by sfan5 — APPROVED** [review]
Works.

**Pull request merged**